### ISSUE TITLE
`GlobalVariable` does not provide a way to determine if it is a builtin or not

### ISSUE THREAD

**Issue by LegNeato** [gfx-rs/wgpu#5417]
**Is your feature request related to a problem? Please describe.**
My [rust backend](https://github.com/gfx-rs/wgpu/pull/5175) needs to translate global variables like `gl_FragCoord` to different names and attributes. The entrypoint attributes work fine, as I can go from `EntryPoint` to `FunctionArgument`s which have a `Binding` that is set to `Binding::BuiltIn(b: BuiltIn)` if the argument is a builtin.

When code is _using_ or referring to a built-in, it appears in the IR as a `GlobalVariable`. Unfortunately, there is no way to go from `GlobalVariable` to `Binding::BuiltIn`. `GlobalVariable` has a binding, but it is a `ResourceBinding` which only has an index and a group.

The problem arises because at _definition time_ the backend gets a `BuiltIn` for the `EntryPoint` `FunctionArg` but there is no name, and at _usage time_ the backend only has the name with no builtin info. It cannot make the two match without   passing around the input format to infer the global name to use as the `EntryPoint` argument name.

**Describe the solution you'd like**
1. Have `Binding` on `GlobalVariable`. I do not know enough about shaders / naga to determine if `Binding` gives all information needed and can replace `ResourceBinding`. 
2. Have the entrypoint argument name not be empty. It should be set to the correct name for the `BuiltIn` in the input format. Then the `Namer` can handle everything.

I _think_ I prefer the second option, but I don't know which is more appropriate. I think just doing the second option means I can have rename rules that rename based on the closed set of GLSL global strings (for example, chop off "gl_"), but I can't generate entirely new names based on the builtin type / have the generated name be frontend-agnostic (again, the calling location only has the name not the builtin's type). Doing both 1 and 2 would provide maximum flexibility and information to the backends.

**Describe alternatives you've considered**
* I think globals will never have a `ResourceBinding` so I can indirectly key off that to determine if it is a builtin? Sadly, I'd have to store a brittle mapping of the string name to the `Builtin` as there would be no way to get the _specific_ builtin.

**Additional context**
Happy to put up a PR if someone points me in the right direction.

**Comment by LegNeato** [comment]
I believe for option 2 above that this needs to be changed:

https://github.com/gfx-rs/wgpu/blob/152a94bc6c502226d9871f28e35db0b755ea35bf/naga/src/front/glsl/variables.rs#L67

I have put up a PR to do so. I tested locally and it works / unblocks me.

I haven't looked at the other frontends. I've been mainly dealing with glsl-in so it could be just a glsl frontend issue. I'll investigate that later.

**Comment by teoxoy** [comment]
The thing is that global variables are never builtins at the IR level. Only the GLSL frontend stores the builtin it gets from the entry point argument into a private variable.

**Comment by teoxoy** [comment]
I don't think there is anything we can do here but feel free to reopen if I misunderstood.

**Comment by LegNeato** [comment]
So, uh, what do I do @teoxoy ? Apologies as I am new to `naga` and shader programming in general.

My backend will be brittle and sometimes break without something like https://github.com/gfx-rs/wgpu/pull/5418. It solves my issue (not completely, but good enough) and I think it is more correct to use the builtin name for the synthesized endpoint argument? I would love to understand why always putting `None` is more correct when the argument clearly comes from something with a name? 

Additionally, the FE _knows_ a global comes from a builtin and then just drops that info on the floor before the backend gets it. That seems bad? Shouldn't that information be preserved so tools or backends can optionally use it if they want without resorting to potentially stale hardcoded string varname lookup tables? Isn't this a variance that exposes FE-specific semantics to the backend rather than encoding it in the IR? Is that considered a bug or at least undesirable?

Concretely, with the referenced PR my `Namer` can accurately handle renaming and both sides match:

```rust

        for (ep_index, ep) in module.entry_points.iter().enumerate() {
            let ep_name = self.call(&ep.name);
            output.insert(NameKey::EntryPoint(ep_index as _), ep_name);
            for (index, arg) in ep.function.arguments.iter().enumerate() {
                let name = match arg.binding {
                    Some(Binding::BuiltIn(_)) => rename_builtin(arg.name.clone()),
                    _ => arg.name.clone(),
                };
                // NOTE: We explicitly do not use `self.call` or `self.call_or` as builtin names
                // are unique already and we don't want to suffix them.
                output.insert(
                    NameKey::EntryPointArgument(ep_index as _, index as u32),
                    name.unwrap_or_else(|| "param".to_string()),
                );
            }
            for (handle, var) in ep.function.local_variables.iter() {
                let name = self.call_or(&var.name, "local");
                output.insert(NameKey::EntryPointLocal(ep_index as _, handle), name);
            }
        }


        for (handle, var) in module.global_variables.iter() {
            let name = if var.binding.is_none() {
                // Assume no binding = global variable.
                rename_builtin(var.name.clone())
            } else {
                var.name.clone()
            };
            // NOTE: We explicitly do not use `self.call` or `self.call_or` as builtin
            // names are unique already and we don't want to suffix them.
            output.insert(
                NameKey::GlobalVariable(handle),
                name.unwrap_or_else(|| "global".to_string()),
            );
        }
```

Without filiing out the name the first call to `rename_builtin` instead needs to be different hardcoded logic that must then be kept in sync.

**I recognize this issue has shifted from the title, so perhaps a different issue should be filed or it should be renamed?**

**Comment by teoxoy** [comment]
All the input builtins get passed as arguments to the entry point. This is an invariant of the IR. Only GLSL uses globals for builtins and the frontend and backend deal with this difference.

> Isn't this a variance that exposes FE-specific semantics to the backend rather than encoding it in the IR? Is that considered a bug or at least undesirable?

We never change the semantics of the IR based on frontend/backend. The IR and its rules are always the same.

**Comment by LegNeato** [comment]
@teoxoy Ok, I see what you are saying!

So this has morphed. Now I don't care about `GlobalVariable`s, now I just want the builtin arguments to have a name. I'll file a new issue.

**Comment by teoxoy** [comment]
I see you are also assuming global variables that don't have a binding to be builtins which is not the case. That might only work for code currently generated by the GLSL frontend but not in all cases (there can be global variables in the private address space i.e. user defined global variables).

**Comment by LegNeato** [comment]
> I see you are also assuming global variables that don't have a binding to be builtins which is not the case. That might only work for code currently generated by the GLSL frontend but not in all cases (there can be global variables in the private address space i.e. user defined global variables).

Thanks for pointing that out, I did not know that!
